test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java

```diff
@@ -0,0 +1,228 @@
+            }
+        });
+
+        assertEquals(ex.getMessage(), INVALID_CEN_BAD_OFFSET);
+    }
+
+    /**
+     * Create an ZIP file with a single entry, then modify the CEN size
+     * in the End of central directory record to the given size.
+     *
+     * The CEN is optionally "inflated" with trailing zero bytes such that
+     * its actual size matches the one stated in the Eoc record.
+     *
+     * The CEN offset is optiontially adjusted by the given amount
+     *
+     * The resulting ZIP is technically not valid, but it does allow us
+     * to test that large or invalid CEN sizes are rejected
+     * @param cenSize the CEN size to put in the END record
+     * @param inflateCen if true, zero-pad the CEN to the desired size
+     * @param cenOffAdjust Adjust the CEN offset field of the END record with this amount
+     */
+    private Path zipWithModifiedEndRecord(int cenSize,
+                                          boolean inflateCen,
+                                          int cenOffAdjust,
+                                          Path zip) throws IOException {
+
+        Files.deleteIfExists(zip);
+
+        // A byte buffer for reading the EOC
+        ByteBuffer buffer = ByteBuffer.wrap(Arrays.copyOf(zipBytes, zipBytes.length))
+                .order(ByteOrder.LITTLE_ENDIAN);
+
+        // Offset of the EOC record
+        int eocOff = buffer.limit() - ENDHDR;
+
+        // Modify the CEN size
+        int sizeOffset = eocOff + ENDSIZ;
+        int currentCenSize = buffer.getInt(sizeOffset);
+        buffer.putInt(sizeOffset, cenSize);
+
+        // Optionally modify the CEN offset
+        if (cenOffAdjust != 0) {
+            int offOffset = eocOff + ENDOFF;
+            int currentCenOff = buffer.getInt(offOffset);
+            buffer.putInt(offOffset, currentCenOff + cenOffAdjust);
+        }
+
+        // Open a FileChannel for writing a sparse file
+        EnumSet<StandardOpenOption> options = EnumSet.of(StandardOpenOption.CREATE_NEW,
+                StandardOpenOption.WRITE,
+                StandardOpenOption.SPARSE);
+
+        try (FileChannel channel = FileChannel.open(zip, options)) {
+
+            // Write everything up to EOC
+            channel.write(buffer.slice(0, buffer.limit() - ENDHDR));
+
+            if (inflateCen) {
+                // Inject "empty bytes" to make the actual CEN size match the EOC
+                int injectBytes = cenSize - currentCenSize;
+                channel.position(channel.position() + injectBytes);
+            }
+            // Write the modified EOC
+            channel.write(buffer.slice(buffer.limit()-ENDHDR, ENDHDR));
+        }
+        return zip;
+    }
+
+    // Produce a byte array of a ZIP with a single entry
+    private byte[] templateZip() throws IOException {
+        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        try (ZipOutputStream zout = new ZipOutputStream(bout)) {
+            zout.putNextEntry(new ZipEntry("duke.txt"));
```

Comment:
Perhaps clarify why you are not writing out any data...

Reply:
No particular reason, so I opted to output some data instead of explaining.